src/angular/components/change-password.component.ts

```diff
@@ -0,0 +1,267 @@
+            return;
+        }
+
+        if (strengthResult != null && strengthResult.score < 3) {
+            const result = await this.platformUtilsService.showDialog(this.i18nService.t('weakMasterPasswordDesc'),
+                this.i18nService.t('weakMasterPassword'), this.i18nService.t('yes'), this.i18nService.t('no'),
+                'warning');
+            if (!result) {
+                return;
+            }
+        }
+
+        if (this.rotateEncKey) {
+            await this.syncService.fullSync(true);
+        }
+
+        const email = await this.userService.getEmail();
+        const kdf = await this.userService.getKdf();
+        const kdfIterations = await this.userService.getKdfIterations();
+        const newKey = await this.cryptoService.makeKey(this.newMasterPassword, email.trim().toLowerCase(),
+            kdf, kdfIterations);
+        const newMasterPasswordHash = await this.cryptoService.hashPassword(this.newMasterPassword, newKey);
+        const newEncKey = await this.cryptoService.remakeEncKey(newKey);
+
+        if (this.isChangePasswordNoCompare) {
+            const setRequest = new SetPasswordRequest();
+            setRequest.newMasterPasswordHash = newMasterPasswordHash;
+            setRequest.key = newEncKey[1].encryptedString;
+
+            try {
+                this.formPromise = this.apiService.setPassword(setRequest);
+                await this.formPromise;
+
+                if (this.onSuccessfulChangePassword != null) {
+                    this.onSuccessfulChangePassword();
+                } else {
+                    this.router.navigate([this.successRoute]);
+                }
+            } catch {
+                this.platformUtilsService.showToast('error', null, this.i18nService.t('errorOccurred'));
+            }
+        } else {
+            const request = new PasswordRequest();
+            request.masterPasswordHash = await this.cryptoService.hashPassword(this.currentMasterPassword, null);
+            request.newMasterPasswordHash = newMasterPasswordHash;
+            request.key = newEncKey[1].encryptedString;
+
+            try {
+                if (this.rotateEncKey) {
+                    this.formPromise = this.apiService.postPassword(request).then(() => {
+                        return this.updateKey(newKey, request.newMasterPasswordHash);
+                    });
+                } else {
+                    this.formPromise = this.apiService.postPassword(request);
+                }
+
+                await this.formPromise;
+
+                this.platformUtilsService.showToast('success', this.i18nService.t('masterPasswordChanged'),
+                    this.i18nService.t('logBackIn'));
+                this.messagingService.send('logout');
+            } catch {
+                this.platformUtilsService.showToast('error', null, this.i18nService.t('errorOccurred'));
+            }
+        }
```

Comment:
I would abstract this into a function that is overwritten for each implementation of this component.

Reply:
Refactored and updated.